src/integrators/batchNormNN.jl

```diff
@@ -0,0 +1,190 @@
+export batchNormNN,getbatchNormNN,initTheta
+
+"""
+batchNormNN Neural Network block
+
+ Y_k+1 = layer{k}(theta{k},Y_k)
+"""
+#TODO: Can probably optimize some functions using the knowledge that we only have a norm and an AFS layer. 
+mutable struct batchNormNN{T,TQ <: Union{Array{T,2},UniformScaling{Int}}} <: AbstractMeganetElement{T}
+    layers   ::Tuple{normLayer{T}, AffineScalingLayer{T}}
+    outTimes ::Array{Int,1}
+    Q        ::TQ
+end
+
+function getbatchNormNN(layers::Tuple{normLayer{T}, AffineScalingLayer{T}},outTimes=eye(Int,length(layers))[:,end],Q=I) where {T <: Number}
+	nt   = length(layers)
+    nout = nFeatOut(layers[1])
+
+    for k=2:nt
+        if nFeatIn(layers[k]) != nout
+            error("Dim. of input features of block $k does not match dim. of output features of block $(k-1)");
+        end
+        nout = nFeatOut(layers[k])
+    end
+	return batchNormNN(layers,outTimes,Q);
+end
+
+# ---------- counting thetas, input and output features -----
+function nTheta(this::batchNormNN)
+    n::Int = 0;
+    for k=1:length(this.layers)
+        n += nTheta(this.layers[k]);
+    end
+    return n
+end
+nFeatIn(this::batchNormNN)   = nFeatIn(this.layers[1])
+nFeatOut(this::batchNormNN) = nFeatOut(this.layers[end])
+
+function nDataOut(this::batchNormNN)
+    n=0;
+    for k=1:length(this.layers)
+        n = n+this.outTimes[k]* nFeatOut(this.layers[k]);
+    end
+end
+
+function initTheta(this::batchNormNN{T}) where {T <: Number}
+    theta = zeros(T,0)
+    for k=1:length(this.layers)
+        theta = [theta; vec(initTheta(this.layers[k]))]
+    end
+    return convert(Array{T},theta)
+end
+
+
+# --------- forward problem ----------
+function apply(this::batchNormNN{T},theta::Array{T},Y0::Array{T,2},doDerivative=true) where {T<:Number}
+    Y::Array{T,2}  = copy(Y0)
+    nex = div(length(Y),nFeatIn(this))::Int
+    nt = length(this.layers)
+
+    tmp = Array{Any}(nt+1,2)
+    if doDerivative
+        tmp[1,1] = Y0
+    end
+
+    Ydata::Array{T,2} = zeros(T,0,nex)
```

Comment:
Instead of initializing with 0 rows, you could init it with with how many non zero entries there are in `this.outTimes`, then the data is pre allocated and you just need to store into it